### ISSUE TITLE
Cannot retrieve server's SSL certificate using custom endpoint factory

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#9843]
|<img alt="michalp2213's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| michalp2213 reported|
|-|-|
|Trac ID|trac#9843|
|Type|defect|
|Created|2020-05-31 20:22:16Z|

What I was trying to do was get server's certificate info when connecting using ```twisted.web.client.Agent```. My approach at this was to create a custom endpoint factory and endpoint classes. My endpoint class is simply a wrapper for the object returned by ```_StandardEndpointFactory.endpointForURI```. It's ```connect``` method adds a callback that is supposed to extract the certificate to the returned deferred. The callback simply calls ```protocol.transport.getPeerCertificate()```, which unexpectedly returns ```None```.

Here is the code to reproduce the behaviour:
```
from twisted.internet import reactor
from twisted.internet.protocol import Factory
from twisted.internet.interfaces import IStreamClientEndpoint
from twisted.web.client import URI, BrowserLikePolicyForHTTPS, _StandardEndpointFactory
from twisted.web._newclient import HTTP11ClientProtocol
from zope.interface import implementer

class CertCaptureEndpointFactory(_StandardEndpointFactory):

    @implementer(IStreamClientEndpoint)
    class CertCaptureEndpointWrapper:
        def __init__(self, endpoint):
            self._endpoint = endpoint

        def connect(self, protocolFactory):
            d = self._endpoint.connect(protocolFactory)
            d.addCallback(self._capture_cert)
            return d

        def _capture_cert(self, protocol):
            print("Certificate: ", protocol.transport.getPeerCertificate())
            return protocol

    def __init__(self, reactor, contextFactory, connectTimeout, bindAddress):
        _StandardEndpointFactory.__init__(self, reactor, contextFactory, connectTimeout, bindAddress)

    def endpointForURI(self, uri):
        endpoint = _StandardEndpointFactory.endpointForURI(self, uri)
        return self.CertCaptureEndpointWrapper(endpoint)


factory = CertCaptureEndpointFactory(
    reactor=reactor,
    contextFactory=BrowserLikePolicyForHTTPS(),
    connectTimeout=10,
    bindAddress=None
)

endpoint = factory.endpointForURI(URI.fromBytes(b'https://example.com'))
d = endpoint.connect(Factory.forProtocol(HTTP11ClientProtocol))
def cbShutdown(ignored):
    reactor.stop()
d.addBoth(cbShutdown)

reactor.run()

```

python version: 3.8.3\\
twisted version: 20.3.0 (installed with pip)

<details><summary>Searchable metadata</summary>

```
trac-id__9843 9843
type__defect defect
reporter__michalp2213 michalp2213
priority__normal normal
milestone__None None
branch__ 
branch_author__ 
status__new new
resolution__None None
component__web web
keywords__None None
time__1590956536311637 1590956536311637
changetime__1590992468682159 1590992468682159
version__None None
owner__None None

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="Gallaecio's avatar" src="https://avatars.githubusercontent.com/u/705211?s=50" width="50" height="50">](https://github.com/Gallaecio)<a name="note_1"></a>|@Gallaecio commented|
|-|-|

Source Scrapy issue, for reference: https://github.com/scrapy/scrapy/issues/4466

**Comment by Renlor** [comment]
I have run into this issue trying to get certificate information from the tls handshake before any data is sent to or received from the remote server.  Here is a reasonably minimal test case.

```python

def main(reactor, host, port=443):
    certificateOptions = OpenSSLCertificateOptions(verify=False)
    options = ClientTLSOptions(host, certificateOptions.getContext())

    port = int(port)

    return endpoints.connectProtocol(
        endpoints.SSL4ClientEndpoint(reactor, host, port, options), ProcessCertificate()
    ).addCallback(lambda protocol: protocol.done)


class ProcessCertificate(protocol.Protocol):
    def __init__(self) -> None:
        super().__init__()

    def connectionMade(self):
        certificate = ssl.Certificate(self.transport.getPeerCertificate())
        if certificate is None:
            print("Twisted bug? Transport did not provide a certificate."
        else:
            print(certificate.dumpPEM())

    def dataReceived(self, data):
        print(data)
        self.transport.abortConnection()

    def connectionLost(self, reason):
        print("Lost.")
       # This one works, but have to wait for time-out
        certificate = ssl.Certificate(self.transport.getPeerCertificate())
        print("Cert:", certificate.dumpPEM())
        self.done.callback(None)


if __name__ == "__main__":
    task.react(main, sys.argv[1:])
```

**Comment by glyph** [comment]
I believe that the issue here is a timing problem. `connectionMade` is called too early. If you wait for `dataReceived`, you should be able to manage it; or if you fudge it with a `callLater(0.1, ...)` you'll probably get the peer cert ready 99% of the time if not 100.

can you give that a try, and see if just waiting a second or so resolves this? If so, I think the solution ought to be straightforward

**Comment by Renlor** [comment]
@glyph Thanks for the suggestion.  Delaying the call does work.  This behavior seems odd to me, isn't connectionMade supposed to be called after the connection is completed?

Is there some technical reason the transport can't be ready before connectionMade must be called or is this just a timing thing in twisted's internals?

I would prefer to wire up an event to when the transport has the certificate set rather than just setting an arbitrary delay and hoping it is there since I need to perform operations using the certificate before and send data before dataReceived will ever be called.

Any suggestions on how I might go about accomplishing this or maybe an interface I need to implement my own version of in Twisted to work with the certificate?  I am not familiar enough with Twisted yet to really know where to look/start, I guess I could build my own implementation of something like IPolicyForHTTPS.

<details>
  <summary>Test Code</summary>

```python

#!/usr/bin/env python3
import sys

from twisted.internet import defer, ssl, task, endpoints, protocol
from twisted.internet._sslverify import OpenSSLCertificateOptions, ClientTLSOptions
from twisted.internet.protocol import connectionDone
from twisted.python.failure import Failure


def main(reactor, host, port=443):
    certificate_options = OpenSSLCertificateOptions(verify=False)
    options = ClientTLSOptions(host, certificate_options.getContext())

    port = int(port)

    return endpoints.connectProtocol(
        endpoints.SSL4ClientEndpoint(reactor, host, port, options), ProcessCertificate(reactor)
    ).addCallback(lambda p: p.done)


class ProcessCertificate(protocol.Protocol):

    def __init__(self, reactor) -> None:
        super().__init__()
        self.reactor = reactor

    def connectionMade(self):
        print("Connection Made Called")
        self.reactor.callLater(1, self.check_certificate)
        self.done = defer.Deferred()
        print("Connection Made Complete")

    def check_certificate(self):
        print("Check Certificate Called")
        certificate = ssl.Certificate(self.transport.getPeerCertificate())
        if certificate is None:
            print("Twisted bug? Transport did not provide a certificate.")
        else:
            print(certificate.dumpPEM())
        self.transport.loseConnection()

    def dataReceived(self, data):
        print(data)
        self.transport.abortConnection()

    def connectionLost(self, reason: Failure = connectionDone) -> None:
        print("Connection Lost Called")
        # This one works, but have to wait for time-out
        certificate = ssl.Certificate(self.transport.getPeerCertificate())
        print("Cert:", certificate.dumpPEM())
        self.done.callback(None)
        print("Connection Lost Complete")


if __name__ == "__main__":
    task.react(main, sys.argv[1:])
```

</details>

**Comment by exarkun** [comment]
If your protocol implements `twisted.internet.interfaces.IHandshakeListener` then you can wait for `handshakeCompleted` instead of `connectionMade`.

It's a bit unfortunate that this separate interface exists but if `connectionMade` (note - actually `makeConnection`) were delayed until the handshake completed, application code would have no way to impose a timeout on how long the handshake takes (since application code wouldn't even know about the connection until the handshake completed).

**Comment by glyph** [comment]
1. we should double-check that this works through a wrapper endpoint, although I'm pretty sure it does now that I am reminded how this works
2. we could probably at least emit a useful warning (maybe even an exception?) here if the user calls `getPeerCertificate`, explaining the availability of the other method

**Comment by adiroiban** [comment]
I think there is still some work to be done here.
Our [howto ssl Connecting To Public Servers](https://docs.twisted.org/en/latest/core/howto/ssl.html#connecting-to-public-servers) should be updated to use the `IHandshakeListener` in that example.

I can see in #8973 that `IHandshakeListener` is expected to work with `SSL4ClientEndpoint`

I am using this example based on `check_server_certificate.py` from the docs

```python
import sys

from zope.interface import implementer
from twisted.internet import defer, endpoints, error, interfaces, protocol, ssl, task

def main(reactor, host, port=443):
    options = ssl.optionsForClientTLS(host)
    return endpoints.connectProtocol(
        endpoints.SSL4ClientEndpoint(reactor, host, int(port), options), ProcessCertificate()
    ).addCallback(lambda protocol: protocol.done)


@implementer(interfaces.IHandshakeListener)
class ProcessCertificate(protocol.Protocol):
    def __init__(self) -> None:
        # Called when connection is lost.
        self.done = defer.Deferred()
        super().__init__()

    def connectionMade(self):
        print('Just connected. Handshake not yet done.')
        # We can send data, even before the TLS hadnshake.
        # It will be buffered, and sent once handshake is done
        # and the TLS connection is ready for data exchange.
        self.transport.write(b"GET / HTTP/1.0\r\n\r\n")
        # But if the connection is closed before handshake is done,
        # it will not wait for handshake, and will close the connection
        # right away.
        # Uncomment the following line to check this behaviour.
        #self.transport.abortConnection()

    def handshakeCompleted(self):
        certificate = self.transport.getPeerCertificate()
        print("Handshake done: {}".format(certificate))

    def dataReceived(self, data):
        print("Received data: {} ...".format(data[:100]))
        self.transport.abortConnection()

    def connectionLost(self, reason):
        handshakeDone = self.transport.getPeerCertificate() != None
        print("\nLost. Handshake done: {}".format(handshakeDone))
        if not reason.check(error.ConnectionClosed):
            print("BAD:", reason.value)
        self.done.callback(None)


task.react(main, sys.argv[1:])
```

If I use it a get

```
python get_cert.py google.com
Just connected. Handshake not yet done.
Handshake done: <OpenSSL.crypto.X509 object at 0x7f6e577a3160>
Received data: b'HTTP/1.0 200 OK\r\nDate: Fri, 29 Sep 2023 08:49:31 GMT\r\nExpires: -1\r\nCache-Control: private, max-age=0' ...

Lost. Handshake done: True
```

**Comment by adiroiban** [comment]
So I guess that this issue can be fixed, just by updating the documentation.

I think it would be nice that instead of protocol. we can have `ProtocolOverTLS` or something like that, that automatically will declare the implementation of `IHandshakeListener`... but this can be controversial.

**Comment by adiroiban** [comment]
@Renlor would you have time to look into the current TLS documentation and example and submit a PR covering the `IHandshakeListener` interface?

Thanks

**Comment by glyph** [comment]
> I think there is still some work to be done here.

Yep, agreed. This seems like a good direction!

**Comment by adiroiban** [comment]
For reference, here is the server-side example

Maybe this can be merged with `echoserv_ssl.py` and have an EchoTLS protocol

```python
import datetime

from cryptography import x509
from cryptography.x509.oid import NameOID
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import rsa
from zope.interface import implementer

from twisted.internet import defer, endpoints, interfaces, protocol, ssl, task
from twisted.python import log
from twisted.python.modules import getModule

# Generate our key
key = rsa.generate_private_key(
    public_exponent=65537,
    key_size=2048,
)

subject = issuer = x509.Name([
    x509.NameAttribute(NameOID.COMMON_NAME, u"DON'T USE IN PROD"),
])
cert = x509.CertificateBuilder().subject_name(
    subject
).issuer_name(
    issuer
).public_key(
    key.public_key()
).serial_number(
    x509.random_serial_number()
).not_valid_before(
    datetime.datetime.now(datetime.timezone.utc)
).not_valid_after(
    # Our certificate will be valid for 10 days
    datetime.datetime.now(datetime.timezone.utc) + datetime.timedelta(days=10)
).add_extension(
    x509.SubjectAlternativeName([x509.DNSName(u"localhost")]),
    critical=False,
# Sign our certificate with our private key
).sign(key, hashes.SHA256())

with open("server.pem", "wb") as f:
    f.write(key.private_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PrivateFormat.TraditionalOpenSSL,
        encryption_algorithm=serialization.NoEncryption(),
    ))

with open("server.pem", "ab") as f:
    f.write(cert.public_bytes(serialization.Encoding.PEM))


@implementer(interfaces.IHandshakeListener)
class TLSServer(protocol.Protocol):
    def connectionMade(self):
        print("New connection.")

    def handshakeCompleted(self):
        print("Handshake done")

    def connectionLost(self, reason):
        print("Lost")

    def dataReceived(self, line):
        print("received: {}".format(line))
        self.transport.write(b'Hello from server.\r\n')


def main(reactor):
    certData = getModule(__name__).filePath.sibling("server.pem").getContent()
    certificate = ssl.PrivateCertificate.loadPEM(certData)
    factory = protocol.Factory.forProtocol(TLSServer)
    reactor.listenSSL(8000, factory, certificate.options())
    return defer.Deferred()

task.react(main)

```

It can be tested using

```
$ echo "Hello from client" | openssl s_client  -connect localhost:8000 -ign_eof
```

**Comment by Renlor** [comment]
@adiroiban I am willing to make an attempt at updating the docs, might be a few weeks.
